### PR TITLE
feat: add `string/to-well-formed`

### LINKED ISSUE #1066: [RFC]: add `@stdlib/string/to-well-formed`
### Description

This RFC proposes adding support for replacing lone surrogates in a provided string in order to create a string which is "well-formed".

This is a functional API analog of the recently standardized `String.prototype.toWellFormed`.

Package: `@stdlib/string/to-well-formed`
alias: `str2wellformed`

### Related Issues

None.

### Questions

No.

### Other

Resources:

- [`String.prototype.toWellFormed`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toWellFormed)

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Pratik772846**
Returns a well-formed string by replacing all lone surrogates

Fixes #1066

Resolves # 1066

## Description

> What is the purpose of this pull request?

This pull request:

-   Replaces all lone surrogates in a string and returns a well-formed string
-   If the string is already well formed then return a shallow copy of the string

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves # 
-   fixes #

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/string/to-well-formed/README.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+
+# str2wellformed
+
+> Creates a new well-formed string.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/README.md]:
```suggestion
> Create a new well-formed string.
```

File: lib/node_modules/@stdlib/string/to-well-formed/README.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+</section>
+
+<!-- /.examples -->
+
+<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->
+
+<section class="related">
+
+
+</section>
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/README.md]: We can put some related packages here.

File: lib/node_modules/@stdlib/string/to-well-formed/benchmark/benchmark.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+var bench = require( '@stdlib/bench' );
+var isString = require( '@stdlib/assert/is-string' ).isPrimitive;
+var pkg = require( './../package.json' ).name;
+var str2wellformed = require( './../lib' );
+
+
+// MAIN //
+
+bench( pkg+'::str2wellformed', function benchmark( b ) {
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/benchmark/benchmark.js]:
```suggestion
bench( pkg, function benchmark( b ) {
```
We don't need this here, any specific reason for having this?

File: lib/node_modules/@stdlib/string/to-well-formed/docs/types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+// TypeScript Version: 4.1
+
+/**
+* Tests if a string is well-formed.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/docs/types/index.d.ts]: Incorrect description

File: lib/node_modules/@stdlib/string/to-well-formed/docs/types/test.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+import str2wellformed = require( './index' );
+
+// TESTS //
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/docs/types/test.ts:22]:
```suggestion


// TESTS //
```

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/string/to-well-formed/docs/types/test.ts:22]: Mek sure you have correct dev env setup

File: lib/node_modules/@stdlib/string/to-well-formed/examples/index.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+*/
+
+/* eslint-disable no-new-wrappers */
+
+'use strict';
+
+var str2wellformed = require( './../lib' );
+
+console.log( str2wellformed( '' ) );
+
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/examples/index.js]:
```suggestion
console.log( str2wellformed( '' ) );
// => ''

```

I will prefer to do it for all console.log here

File: lib/node_modules/@stdlib/string/to-well-formed/lib/index.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+* // returns ab�
+*
+* str = "ab\uD83D\uDE04c";
+* newStr = str2wellformed( str );
+* // returns ab😄c
+*
+* str = "abc";
+* bool = str2wellformed( str );
+* // returns abc
+*
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/lib/index.js]:
```suggestion
```

File: lib/node_modules/@stdlib/string/to-well-formed/lib/main.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+* @throws {TypeError} must provide a string
+* @returns {string} new string which does not contain any lone surrogates
+*
+* @example
+* var str = str2wellformed( 'ab\uD800' );
+* // returns 'ab�'
+*/
+function str2wellformed( str ) {
+	var modifiedString = '';
+	var i;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/lib/main.js]:
```suggestion
	var modifiedString;
	var i;
	modifiedString = '';
```

File: lib/node_modules/@stdlib/string/to-well-formed/lib/main.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	if ( !isString( str ) ) {
+		throw new TypeError( format( 'invalid argument. Must provide a string. Value: `%s`.', str ) );
+	}
+	if ( isWellFormed( str ) ) {
+		for ( i = 0; i < str.length; i++ ) {
+			modifiedString += str[i];
+		}
+		return modifiedString;
+	}
+	for ( i =0; i < str.length; i++ ) {
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/lib/main.js]:
```suggestion
	for ( i = 0; i < str.length; i++ ) {
```

File: lib/node_modules/@stdlib/string/to-well-formed/lib/main.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+		if ( RE_UTF16_LOW_SURROGATE.test( str[i] ) || RE_UTF16_HIGH_SURROGATE.test( str[i] ) ) {
+			modifiedString += '\uFFFD';
+		}
+		else {
+			modifiedString += str[i];
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/lib/main.js]:
```suggestion
			modifiedString += str[ i ];
```

File: lib/node_modules/@stdlib/string/to-well-formed/package.json [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+      "negative",
+      "is",
+      "isnumber",
+      "isnumeric",
+      "type",
+      "check",
+      "primitive",
+      "object"
+    ]
+  }
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/package.json]: incorrect indentation

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Thanks @Pratik772846 for working on this. Requested changes.

File: lib/node_modules/@stdlib/string/to-well-formed/package.json [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    "is",
+    "isnumber",
+    "isnumeric",
+    "type",
+    "check",
+    "primitive",
+    "object"
+  ]
+}
+
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/string/to-well-formed/package.json]:
```suggestion
```

**Review by Pranavchiku — APPROVED** [review]
LGTM.

**Comment by Pranavchiku** [comment]
I am not so great in reviewing this kind of PRs, I let Athan or Phillip review this.

**Comment by Pratik772846** [comment]
@Planeshifter Can you please review these changes?

File: lib/node_modules/@stdlib/string/to-well-formed/docs/repl.txt [hunk 13]
```diff
@@ -0,0 +1,31 @@
+
+{{alias}}( str )
+    Returns a well formed string.
+
+    Parameters
+    ----------
+    str: string
+        String to test.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/to-well-formed/docs/repl.txt]:
```suggestion
        Input string.
```

File: lib/node_modules/@stdlib/string/to-well-formed/package.json [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    "number",
+    "numeric",
+    "negative",
+    "is",
+    "isnumber",
+    "isnumeric",
+    "type",
+    "check",
+    "primitive",
+    "object"
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/to-well-formed/package.json]:
```suggestion
    "stdlib",
    "string",
    "well-formed",
    "surrogate",
    "lone-surrogate",
    "utf-16",
    "unicode",
    "replacement",
    "character",
    "validation",
    "sanitization",
    "conversion",
    "transform",
    "utility",
    "utils"
```

File: lib/node_modules/@stdlib/string/to-well-formed/test/test.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+tape( 'the function returns a well-formed string by replacing lone surrogates with Unicode replacement character', function test( t ) {
+	t.equal(str2wellformed(''), '', 'returns an empty string for an empty input');
+	t.equal(str2wellformed('\uDBFF'), '�', 'replaces lone high surrogate with Unicode replacement character');
+	t.equal(str2wellformed('\uDBFFFF\uDBFF'), '�FF�', 'replaces multiple lone surrogates with Unicode replacement characters');
+	t.equal(str2wellformed('Hello \uDBFFWorld'), 'Hello �World', 'replaces lone high surrogate in a string with Unicode replacement character');
+	t.equal(str2wellformed('\uDBFF'), '�', 'replaces lone high surrogate with Unicode replacement character');
+	t.equal(str2wellformed('\uDBFFFFF'), '�FFF', 'replaces lone high surrogate with Unicode replacement character');
+	t.equal(str2wellformed('\uDC00\uDC00'), '��', 'replaces lone low surrogate with Unicode replacement character');
+	t.equal(str2wellformed('5\uDBFF'), '5�', 'replaces lone high surrogate with Unicode replacement character');
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/to-well-formed/test/test.js]:
```suggestion
	t.equal( str2wellformed(''), '', 'returns an empty string for an empty input' );
	t.equal( str2wellformed('\uDBFF'), '�', 'replaces lone high surrogate with Unicode replacement character' );
	t.equal( str2wellformed('\uDBFFFF\uDBFF'), '�FF�', 'replaces multiple lone surrogates with Unicode replacement characters' );
	t.equal( str2wellformed('Hello \uDBFFWorld'), 'Hello �World', 'replaces lone high surrogate in a string with Unicode replacement character' );
	t.equal( str2wellformed('\uDBFF'), '�', 'replaces lone high surrogate with Unicode replacement character' );
	t.equal( str2wellformed('\uDBFFFFF'), '�FFF', 'replaces lone high surrogate with Unicode replacement character' );
	t.equal( str2wellformed('\uDC00\uDC00'), '��', 'replaces lone low surrogate with Unicode replacement character' );
	t.equal( str2wellformed('5\uDBFF'), '5�', 'replaces lone high surrogate with Unicode replacement character' );
```

File: lib/node_modules/@stdlib/string/to-well-formed/package.json [hunk 16]
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "@stdlib/string/to-well-formed",
+  "version": "0.0.0",
+  "description": "Convert a string to a well formed string",
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/to-well-formed/package.json]:
```suggestion
  "description": "Convert a string to a well formed string.",
```

File: lib/node_modules/@stdlib/string/to-well-formed/lib/index.js [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+/**
+* Return a well formed string.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/to-well-formed/lib/index.js]:
```suggestion
* Convert a string to a well formed string.
```

File: lib/node_modules/@stdlib/string/to-well-formed/docs/types/index.d.ts [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+// TypeScript Version: 4.1
+
+/**
+* Returns a well formed string.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/to-well-formed/docs/types/index.d.ts]:
```suggestion
* Replaces all lone surrogates in order to create a new string which is well formed.
```

File: lib/node_modules/@stdlib/string/to-well-formed/docs/types/index.d.ts [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+*
+* @param str - string to test
+* @returns returns a well formed string
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/to-well-formed/docs/types/index.d.ts]:
```suggestion
* @returns new string which does not contain any lone surrogates
```

File: lib/node_modules/@stdlib/string/to-well-formed/README.md [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+## Usage
+
+```javascript
+var str2wellformed = require( '@stdlib/string/to-well-formed' );
+```
+
+#### str2wellformed( str )
+
+Creates a new well-formed string.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/to-well-formed/README.md]:
```suggestion
Creates a new well-formed string by replacing all lone surrogates with the replacement character `\uFFFD`.
```

**Comment by Planeshifter** [comment]
@Pratik772846 Implementation looks fine; will merge after CI has passed. Thank you!

**Pull request merged**